### PR TITLE
Rollup add default metrics to histo groups

### LINKED ISSUE #34379: Add default metrics for groupings in Rollups
In `date_histogram` and `histogram` group bys we should by default include the following metrics on the grouping value:
- `max`
- `min`
- `count`

Discussed with @polyfractal @colings86

### PR DESCRIPTION
**Pull request by benwtrent**
This adds default metrics `[min, max]` to histogram groupings (numeric and date). 

I was not 100% sure to put these default metrics, but keeping them as simple metric aggregations keeps it simple and makes it more obvious to the end user what capabilities are there for what fields. 

If a user provides a metric for a histogram field, it is assumed that they want that specific metric only.

Docs need to be updated as well.

closes #34379

P.S. Would be nice to get this into 6.5, but could be delayed to a next version if need be.

### CODE REVIEW

**Comment by Conky5** [comment]
jenkins test this please

**Review by colings86 — COMMENTED** [review]
@benwtrent This looks good to me but I wonder if we should have a test to check that the override behaviour works as expected (i.e. if the user provides explicit metrics for a histo field their settings prevail)? 

Also I think we should add documentation that tells the user that we will add these metrics by default and also explains the override behaviour?

**Comment by benwtrent** [comment]
@colings86 Yes, I am going to add documentation for it, I will update the HLRC docs in this PR and will open another one for the docs repo. I wanted to make sure that this looked ok before I started working on that. 

The tests cover the default value vs override situation that you describe.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java [hunk 1]
```diff
@@ -128,7 +137,7 @@ public RollupJobConfig(final String id,
         this.indexPattern = indexPattern;
         this.rollupIndex = rollupIndex;
         this.groupConfig = groupConfig;
-        this.metricsConfig = metricsConfig != null ? metricsConfig : Collections.emptyList();
+        this.metricsConfig = addDefaultMetricsIfNeeded(metricsConfig, groupConfig);
```
Comment (polyfractal) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java]: Since we always guarantee at least one metric is included (either explicitly, or through default), I think we can get rid of the `"At least one grouping or metric must be configured"` check a little bit above this line.  That was just to make sure users didn't configure a completely empty Rollup.  

I think that's moot now that we include some defaults, and an empty rollup with those defaults might still be useful (min/max of histos per bucket, etc).

We can also remove the null checks on `metricsConfig` in the `getAllFields()` and `toXContent()` methods I think?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java [hunk 2]
```diff
@@ -47,6 +53,9 @@
     private static final String PAGE_SIZE = "page_size";
     private static final String INDEX_PATTERN = "index_pattern";
     private static final String ROLLUP_INDEX = "rollup_index";
+    private static final List<String> DEFAULT_HISTO_METRICS = Arrays.asList(MaxAggregationBuilder.NAME,
+        MinAggregationBuilder.NAME,
+        ValueCountAggregationBuilder.NAME);
```
Comment (polyfractal) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java]: Do we need `value_count` as a default?  Both histo's naturally save their doc_count (since it's part of the return data of a histo/date_histo agg).  Is there a specific need for value counts instead of just doc count?

File: docs/java-rest/high-level/rollup/put_job.asciidoc [hunk 3]
```diff
@@ -119,6 +119,68 @@ include-tagged::{doc-tests}/RollupDocumentationIT.java[x-pack-rollup-put-rollup-
 <2> Adds the metrics to compute on the `temperature` field
 <3> Adds the metrics to compute on the `voltage` field
 
+`RollupJobConfig` also sets default metrics for `DateHistogram` and `Histogram` groupings.
+If the user does not provide a `MetricConfig` referencing the field in a `DateHistogram` or
```
Comment (polyfractal) [inline docs/java-rest/high-level/rollup/put_job.asciidoc]: Could we reword this slightly so it is shorter sentences? Maybe something like:

>By default, `min`/`max`/`value_count` metrics will be added to the `DateHistogram` and `Histogram`.  This behavior can be overridden by explicitly configuring metrics for the `DateHistogram` and `Histogram` groups.  

Or something.  The "If the user..." sentence just ran a bit long and was hard to parse.

**Review by polyfractal — COMMENTED** [review]
Left some comments/questions.  Just minor stuff, looks good to me.

Not crazy in love with how we need to make these changes in the main config object... but I don't see a better way since it's a List of MetricConfigs rather than an object which can own the defaults.  

>  I will update the HLRC docs in this PR and will open another one for the docs repo

We can do the changes to the REST/reference docs in this PR, no need for a second one.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java (hunk 1 above)
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java]: @polyfractal if both are null, then it will still be an empty rollup as there would be no groupings from which to derive default metrics.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java (hunk 2 above)
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java]: Probably not, I can remove it.

Comment (polyfractal) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java]: Oops sorry, this probably should have a top-level comment in the ticket, not on a source line.  I just meant in general, do we need to default to storing value_count? Min/max make sense, i'm less sold on value count (since we already have doc count)

@colings86 wdyt?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java (hunk 1 above)
Comment (polyfractal) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java]: We enforce that a date_histo is always configured, so at a minimum there will be buckets of time.

We can leave the check for now, it might not be useful to just have time buckets + min/max time.  Probably easier to remove later if a user needs it, rather than allow it and regret it later :)

Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java]: Yes, we require a `date_histo` in the `GroupConfig` object, but if that config is null, then we have nothing from which to gather default values. 

I think if the desire is for a date_histo to ALWAYS be provided, there may be a bug. If `GroupConfig` is null, but a `MetricConfig` objects are provided, then that null check passes and no date_histogram configuration is provided. 

Additionally `validateMappings` would throw a NPE if the groupConfig was null.

Comment (polyfractal) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java]: Yeah I think you're right, that's an NPE waiting to happen.  Let's ignore that for now in this PR to keep the changes minimal... I (or you if you want! ;) ) can followup with a bugfix for that null groupconfig issue.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java (hunk 2 above)
Comment (colings86) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java]: I agree I think the majority of use cases will care about doc count rather than value count so if a user explicitly wants value count I think they can use the override mechanism to add it instead of use storing this extra information

Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/RollupJobConfig.java]: Alright, I will pull `value_count` out of our default metrics. Will update the PR shortly :)

**Review by polyfractal — APPROVED** [review]
Left some comments about the documentation, but otherwise I think this is good to go.  I'm happy with it once the docs are tweaked and CI passes :)

Thanks @benwtrent!

File: docs/reference/rollup/apis/put-job.asciidoc [hunk 4]
```diff
@@ -98,4 +98,81 @@ When the job is created, you receive the following results:
...
+    },
+    "metrics": [
+        {
+            "field": "temperature",
+            "metrics": ["min", "max", "sum"]
+        }
+    ]
+}
+--------------------------------------------------
+// NOTCONSOLE
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply docs/reference/rollup/apis/put-job.asciidoc]: Test clause has been added.
